File: src/domain/schemas/auto-funding.ts
/**
 * Auto-Funding Domain Schemas
 * Runtime validation for auto-funding rule and execution record entities
 */

import { z } from "zod";

/**
 * Condition schema for auto-funding rules
 */
export const ConditionSchema = z.object({
  id: z.string().optional(),
  type: z.string().min(1, "Condition type is required"),
  envelopeId: z.string().nullable().optional(),
  value: z.number(),
  operator: z.string().optional(),
  startDate: z.string().nullable().optional(),
  endDate: z.string().nullable().optional(),
});

export type Condition = z.infer<typeof ConditionSchema>;

/**
 * Auto-funding rule configuration schema
 */
export const AutoFundingConfigSchema = z.object({
  sourceType: z.enum(["unassigned", "envelope", "income"]),
  sourceId: z.string().nullable(),
  targetType: z.enum(["envelope", "multiple"]),
  targetId: z.string().nullable(),
  targetIds: z.array(z.string()),
  amount: z.number().min(0, "Amount cannot be negative"),
  percentage: z.number().min(0).max(100, "Percentage must be between 0 and 100"),
  conditions: z.array(ConditionSchema),
  scheduleConfig: z.record(z.string(), z.unknown()),
});

export type AutoFundingConfig = z.infer<typeof AutoFundingConfigSchema>;

/**
 * Auto-funding rule schema
 */
export const AutoFundingRuleSchema = z.object({
  id: z.string().min(1, "Rule ID is required"),
  name: z
    .string()
    .min(1, "Rule name is required")
    .max(100, "Rule name must be 100 characters or less"),
  description: z.string().max(500, "Description must be 500 characters or less"),
  type: z.string().min(1, "Rule type is required"),
  trigger: z.string().min(1, "Trigger is required"),
  priority: z.number().int().min(0, "Priority must be non-negative"),
  enabled: z.boolean(),
  createdAt: z.string(),
  lastExecuted: z.string().nullable(),
  executionCount: z.number().int().min(0, "Execution count cannot be negative"),
  config: AutoFundingConfigSchema,
  lastModified: z.number().int().positive("Last modified must be a positive number"),
});

export type AutoFundingRule = z.infer<typeof AutoFundingRuleSchema>;

/**
 * Partial auto-funding rule schema for updates
 */
export const AutoFundingRulePartialSchema = AutoFundingRuleSchema.partial();
export type AutoFundingRulePartial = z.infer<typeof AutoFundingRulePartialSchema>;

/**
 * Execution record schema
 */
export const ExecutionRecordSchema = z.object({
  id: z.string().min(1, "Execution record ID is required"),
  trigger: z.string().min(1, "Trigger is required"),
  totalFunded: z.number().optional(),
  success: z.boolean().optional(),
  executedAt: z.string().optional(),
  rulesExecuted: z.number().int().min(0).optional(),
  timestamp: z.string().optional(), // Legacy/Alias support
  lastModified: z.number().int().positive("Last modified must be a positive number"),
});

export type ExecutionRecord = z.infer<typeof ExecutionRecordSchema>;

/**
 * Partial execution record schema for updates
 */
export const ExecutionRecordPartialSchema = ExecutionRecordSchema.partial();
export type ExecutionRecordPartial = z.infer<typeof ExecutionRecordPartialSchema>;

/**
 * Validation helpers
 */
export const validateAutoFundingRule = (data: unknown): AutoFundingRule =>
  AutoFundingRuleSchema.parse(data);

export const validateAutoFundingRuleSafe = (data: unknown) => AutoFundingRuleSchema.safeParse(data);

export const validateAutoFundingRulePartial = (data: unknown): AutoFundingRulePartial =>
  AutoFundingRulePartialSchema.parse(data);

export const validateAutoFundingRulePartialSafe = (data: unknown) =>
  AutoFundingRulePartialSchema.safeParse(data);

export const validateExecutionRecord = (data: unknown): ExecutionRecord =>
  ExecutionRecordSchema.parse(data);

export const validateExecutionRecordSafe = (data: unknown) => ExecutionRecordSchema.safeParse(data);

export const validateExecutionRecordPartial = (data: unknown): ExecutionRecordPartial =>
  ExecutionRecordPartialSchema.parse(data);

export const validateExecutionRecordPartialSafe = (data: unknown) =>
  ExecutionRecordPartialSchema.safeParse(data);

File: api/models/autofunding.go
package models

// Rule represents an autofunding rule
type Rule struct {
	ID             string     `json:"id"`
	Name           string     `json:"name"`
	Description    string     `json:"description,omitempty"`
	Type           string     `json:"type"`
	Trigger        string     `json:"trigger"`
	Priority       int        `json:"priority"`
	Enabled        bool       `json:"enabled"`
	Config         RuleConfig `json:"config"`
	CreatedAt      string     `json:"createdAt,omitempty"`
	LastExecuted   *string    `json:"lastExecuted,omitempty"`
	ExecutionCount int        `json:"executionCount,omitempty"`
}

// RuleConfig contains the configuration for a rule
type RuleConfig struct {
	SourceType     string       `json:"sourceType"`
	SourceID       *string      `json:"sourceId,omitempty"`
	TargetType     string       `json:"targetType"`
	TargetID       *string      `json:"targetId,omitempty"`
	TargetIDs      []string     `json:"targetIds,omitempty"`
	Amount         int64        `json:"amount,omitempty"`
	Percentage     float64      `json:"percentage,omitempty"`
	Conditions     []Condition  `json:"conditions,omitempty"`
	ScheduleConfig *interface{} `json:"scheduleConfig,omitempty"`
}

// Condition represents a conditional check for a rule
type Condition struct {
	Type       string  `json:"type"`
	EnvelopeID *string `json:"envelopeId,omitempty"`
	Value      int64   `json:"value,omitempty"`
	Operator   *string `json:"operator,omitempty"`
}

// Envelope represents an envelope with balance and target
type Envelope struct {
	ID             string `json:"id"`
	Name           string `json:"name,omitempty"`
	CurrentBalance int64  `json:"currentBalance"`
	MonthlyTarget  int64  `json:"monthlyTarget,omitempty"`
	Category       string `json:"category,omitempty"`
}

// AllocationContext provides context for rule execution
type AllocationContext struct {
	UnassignedCash  int64      `json:"unassignedCash"`
	NewIncomeAmount int64      `json:"newIncomeAmount,omitempty"`
	Envelopes       []Envelope `json:"envelopes"`
}

// Allocation represents a fund allocation result
type Allocation struct {
	EnvelopeID  string  `json:"envelopeId"`
	AmountCents int64   `json:"amountCents"`
	Reason      string  `json:"reason,omitempty"`
	RuleID      *string `json:"ruleId,omitempty"`
}

// AutofundingRequest is the request payload for rule execution
type AutofundingRequest struct {
	RuleIDs []string          `json:"ruleIds,omitempty"`
	Trigger string            `json:"trigger"`
	Context AllocationContext `json:"context"`
}

// AutofundingResponse is the response from rule execution
type AutofundingResponse struct {
	Allocations         []Allocation `json:"allocations"`
	TotalAllocatedCents int64        `json:"totalAllocatedCents"`
	RemainingCents      int64        `json:"remainingCents"`
	ExecutionTimeMs     float64      `json:"executionTimeMs"`
	RulesExecuted       int          `json:"rulesExecuted"`
}
